### ISSUE TITLE
🧮 Implement Additional Tensor Operations

### ISSUE THREAD

**Issue by sakshambedi** [sakshambedi/micrograd#16]
# 🧮 Implement Additional Tensor Operations

## 📋 Description

Implement essential tensor operations that are currently missing from the micrograd library. These operations are fundamental for scientific computing, machine learning, and neural network implementations.

## 🎯 Acceptance Criteria

- [ ] Implement `Tensor.mean()` with axis and keepdim support
- [ ] Add reduction operations: `max()`, `min()`, `argmax()`, `argmin()`
- [ ] Implement mathematical functions: `exp()`, `log()`, `sqrt()`, `abs()`
- [ ] Add trigonometric functions: `sin()`, `cos()`, `tan()`
- [ ] Implement activation functions: `relu()`, `sigmoid()`, `tanh()`
- [ ] Add tensor comparison operations: `eq()`, `gt()`, `lt()`, `ge()`, `le()`
- [ ] Implement `where()` for conditional selection
- [ ] Add `clamp()` for value clipping
- [ ] Implement proper gradient computation for all operations
- [ ] Add comprehensive tests for all new operations

### Current State
- `Tensor.sum()`: ✅ Implemented (basic version)
- `Tensor.mean()`: ❌ Stub only
- Math functions: ❌ Not implemented
- Reduction ops: ❌ Not implemented
- Activations: ❌ Not implemented

### Implementation Requirements

**Mean Operation**
```python
@staticmethod
def mean(t: Tensor, axis: Optional[int] = None, keepdim: bool = False) -> Tensor:
    """Compute mean along specified axis"""
    # Implementation with sum() and division
    total = Tensor.sum(t, axis=axis, keepdim=keepdim)
    count = _prod(t.shape) if axis is None else t.shape[axis]
    return total / count
```

**Mathematical Functions**
```python
class Exp(Function):
    @staticmethod
    def forward(ctx, a):
        result = _unary_op(a, lambda x: math.exp(x))
        ctx.save_for_backward(result)
        return result
    
    @staticmethod 
    def backward(ctx, grad_output):
        result, = ctx.saved_tensors
        return result * grad_output  # d/dx(exp(x)) = exp(x)
```

**Activation Functions**
```python
class ReLU(Function):
    @staticmethod
    def forward(ctx, a):
        mask = _unary_op(a, lambda x: x > 0)
        ctx.save_for_backward(mask)
        return _unary_op(a, lambda x: max(0, x))
    
    @staticmethod
    def backward(ctx, grad_output):
        mask, = ctx.saved_tensors
        return mask * grad_output
```

### Operations to Implement

**Reduction Operations**
- [ ] `mean(axis=None, keepdim=False)`
- [ ] `max(axis=None, keepdim=False)` 
- [ ] `min(axis=None, keepdim=False)`
- [ ] `argmax(axis=None)`
- [ ] `argmin(axis=None)`

**Mathematical Functions**
- [ ] `exp()` - Exponential function
- [ ] `log()` - Natural logarithm
- [ ] `sqrt()` - Square root
- [ ] `abs()` - Absolute value
- [ ] `pow(exponent)` - Power with scalar exponent

**Trigonometric Functions**
- [ ] `sin()` - Sine
- [ ] `cos()` - Cosine  
- [ ] `tan()` - Tangent

**Activation Functions**
- [ ] `relu()` - Rectified Linear Unit
- [ ] `sigmoid()` - Sigmoid activation
- [ ] `tanh()` - Hyperbolic tangent

**Comparison Operations**
- [ ] `eq(other)` - Element-wise equality
- [ ] `gt(other)` - Greater than
- [ ] `lt(other)` - Less than
- [ ] `ge(other)` - Greater or equal
- [ ] `le(other)` - Less or equal

**Utility Operations**  
- [ ] `where(condition, x, y)` - Conditional selection
- [ ] `clamp(min_val, max_val)` - Value clipping

## 🧪 Testing Requirements

- [ ] Test all operations with different tensor shapes
- [ ] Test with different data types
- [ ] Verify gradients using numerical differentiation
- [ ] Test edge cases (zeros, infinities, NaN)
- [ ] Test axis parameter behavior for reduction operations
- [ ] Performance benchmarks against NumPy
- [ ] Test integration with existing autograd system

## 📁 Files to Modify

- `grad/tensor.py` - Add method declarations and operator overloads
- `grad/autograd/ops.py` - Implement Function classes for each operation
- `grad/autograd/functional.py` - New file for functional interface
- `tests/test_tensor_ops.py` - Comprehensive test suite
- `tests/test_activations.py` - Activation function tests
- `tests/test_reductions.py` - Reduction operation tests

## 🔗 Dependencies

- Requires: Autograd backward pass (#01)
- Requires: Broadcasting support (#02) 
- Uses: Existing `_unary_op()` and `_elementwise_op()` helpers
- Blocks: Neural network layers (#08)
- Blocks: Loss functions (#09)

### Usage Examples
```python
# Mathematical operations
x = Tensor([1.0, 2.0, 3.0], requires_grad=True)
y = x.exp().log()  # Should equal x
loss = (y - x).abs().mean()
loss.backward()

# Activation functions
hidden = Tensor.randn(64, 128, requires_grad=True)
activated = hidden.relu()
output = activated.sigmoid()

# Reduction operations
data = Tensor.randn(10, 20, 30)
mean_val = data.mean(axis=1, keepdim=True)  # Shape: (10, 1, 30)
max_val = data.max(axis=2)  # Shape: (10, 20)

# Conditional operations
condition = x > 0
result = Tensor.where(condition, x, -x)  # Absolute value

# Clipping
clipped = x.clamp(0, 1)  # Clip to [0, 1] range
```

### Gradient Computation
```python
x = Tensor([0.5], requires_grad=True)
y = x.sin().cos().exp()
y.backward()
print(x.grad)  # Chain rule gradient
```

## 🚨 Edge Cases to Handle

- Division by zero in log operations
- Square root of negative numbers
- Numerical stability for sigmoid/tanh
- Overflow in exponential functions
- NaN and infinity propagation
- Empty tensors in reduction operations
- Invalid axis parameters

## 🔧 Performance Considerations

- Efficient implementation of transcendental functions
- Memory-efficient reduction operations
- Vectorized operations where possible
- Numerical stability optimizations
- Consider future SIMD optimizations

## 🏷️ Labels

- `enhancement`
- `tensor-operations`
- `mathematical-functions`
- `medium-priority`

## 📈 Priority

**Medium** - Important for completeness and neural network implementations

## 🔗 Related Issues

- Depends on: Autograd system (#01, #03)
- Depends on: Broadcasting (#02)
- Blocks: Neural network layers (#08)
- Blocks: Optimization algorithms (#10)
- Related to: Numerical stability improvements
